crates/assistant/src/assistant_panel.rs

```diff
@@ -3121,6 +3198,36 @@ impl ContextEditor {
         }
     }
 
+    fn copy_selection(
+        workspace: &mut Workspace,
+        _: &CopySelection,
+        cx: &mut ViewContext<Workspace>,
+    ) {
+        let Some(panel) = workspace.panel::<AssistantPanel>(cx) else {
+            return;
+        };
+        let Some(context_editor_view) = panel.read(cx).active_context_editor(cx) else {
+            return;
+        };
+
+        let (text, is_code_block) = Self::get_selection_or_code_block(&context_editor_view, cx);
+
+        if !text.is_empty() {
+            cx.write_to_clipboard(ClipboardItem::new_string(text));
+            if is_code_block {
```

Comment:
Nitpick: why do we only show this when it's a code block?

Reply:
> Nitpick: why do we only show this when it's a code block?

I figured because copying while you have a selection is quite normal and you don't expect to get any feedback for it, whereas copying "nothing" is a little non standard. But yeah, maybe the distinction isn't necessary

Reply:
Yeah, I'm a bit unsure about the mechanics. I guess my question is: why do we need a new command if there's a selection? If there's a selection, you can `cmd-c` to copy, right? Should we maybe make your new command a no-op if there is a selection? Or should it expand the selection and then copy?

Reply:
I didn't do a no-op because as I said I want to be consistent with the new insert into editor, which is: insert selection if there is one (old behavior), else insert the code block if inside a code block (new behavior).

Visually selecting the block is a good idea and it's what I tried to do but I couldn't figure out the code. Specifically, I don't know how to get the `cx: &mut ViewContext<Self>` param for `Editor > change_selections` when I have `cx: &mut ViewContext<Workspace>`. Can you help? If I can select the code block, that will make it clear what's happening and we can get rid of the toast message.

Reply:
> I didn't do a no-op because as I said I want to be consistent with the new insert into editor, which is: insert selection if there is one (old behavior), else insert the code block if inside a code block (new behavior).

Ah, I see. I didn't realise that you also modified the existing action. Sorry!



> Specifically, I don't know how to get the `cx: &mut ViewContext<Self>` param for `Editor > change_selections` when I have `cx: &mut ViewContext<Workspace>`. Can you help? If I can select the code block, that will make it clear what's happening and we can get rid of the toast message.

Sure! The `context_editor_view.read(cx).editor` is a `View<Editor>`. If you call `update` on that, you can pass in a callback that receives `editor` and `cx`, which is the `ViewContext<Editor>`. Like this:

```rust
fn get_selection_or_code_block(
    context_editor_view: &View<ContextEditor>,
    cx: &mut ViewContext<Workspace>,
) -> (String, bool) {
    // (you don't need to specify the types, I'm just doing it here for illustration)
    context_editor_view.read(cx).editor.update(cx, |editor: &mut Editor, cx: &mut ViewContext<Editor>| {
        
    });
    // ...
}
```

Reply:
thanks. I needed to throw in a clone() after read(cx).editor and rearrange the code a bit to make the borrow checker happy, but it worked. Will push an update later today :)